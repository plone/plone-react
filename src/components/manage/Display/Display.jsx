/**
 * Display component.
 * @module components/manage/Display/Display
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Dropdown, Icon } from 'semantic-ui-react';
import { FormattedMessage } from 'react-intl';

import { getSchema, editContent, getContent } from '../../../actions';
import layouts from '../../../constants/Layouts';

/**
 * Display container class.
 * @class Display
 * @extends Component
 */
@connect(
  state => ({
    loaded: state.content.edit.loaded,
    layouts: state.schema.schema ? state.schema.schema.layouts : [],
    layout: state.content.data ? state.content.data.layout : '',
    type: state.content.data ? state.content.data['@type'] : '',
  }),
  dispatch =>
    bindActionCreators({ getSchema, editContent, getContent }, dispatch),
)
export default class Display extends Component {
  /**
   * Property types.
   * @property {Object} propTypes Property types.
   * @static
   */
  static propTypes = {
    getSchema: PropTypes.func.isRequired,
    editContent: PropTypes.func.isRequired,
    getContent: PropTypes.func.isRequired,
    loaded: PropTypes.bool.isRequired,
    pathname: PropTypes.string.isRequired,
    layouts: PropTypes.arrayOf(PropTypes.string),
    layout: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
    expanded: PropTypes.bool,
  };

  /**
   * Default properties
   * @property {Object} defaultProps Default properties.
   * @static
   */
  static defaultProps = {
    history: [],
    transitions: [],
    layouts: [],
    expanded: true,
  };

  /**
   * Constructor
   * @method constructor
   * @param {Object} props Component properties
   * @constructs Workflow
   */
  constructor(props) {
    super(props);
    this.setLayout = this.setLayout.bind(this);
  }

  /**
   * Component will mount
   * @method componentWillMount
   * 
   */
  componentWillMount() {
    this.props.getSchema(this.props.type);
  }

  /**
   * Component will receive props
   * @method componentWillReceiveProps
   * @param {Object} nextProps Next properties
   * 
   */
  componentWillReceiveProps(nextProps) {
    if (nextProps.pathname !== this.props.pathname) {
      this.props.getSchema(nextProps.type);
    }
    if (!this.props.loaded && nextProps.loaded) {
      this.props.getContent(nextProps.pathname);
    }
  }

  /**
   * On set layout handler
   * @method setLayout
   * @param {Object} event Event object
   * 
   */
  setLayout(event, { value }) {
    this.props.editContent(this.props.pathname, {
      layout: value,
    });
  }

  /**
   * Render method.
   * @method render
   * @returns {string} Markup for the component.
   */
  render() {
    return (
      <Dropdown
        item
        trigger={
          <span>
            <Icon name="block layout" />{' '}
            {this.props.expanded && (
              <FormattedMessage id="Display" defaultMessage="Display" />
            )}
          </span>
        }
        pointing="left"
      >
        <Dropdown.Menu>
          {this.props.layouts.map(item => (
            <Dropdown.Item
              text={layouts[item] || item}
              value={item}
              active={this.props.layout === item}
              key={item}
              onClick={this.setLayout}
            />
          ))}
        </Dropdown.Menu>
      </Dropdown>
    );
  }
}
