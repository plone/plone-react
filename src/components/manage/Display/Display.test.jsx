import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-intl-redux';
import { wait } from '@testing-library/react';

import Display from './Display';

const mockStore = configureStore();

describe('Display', () => {
  it('renders an actions component', async () => {
    const store = mockStore({
      content: {
        update: { loaded: true },
        data: { layout: 'summary_view', '@type': 'Folder' },
      },
      schema: {
        schema: {
          layouts: ['summary_view'],
        },
      },
      intl: {
        locale: 'en',
        messages: {},
      },
    });
    const component = renderer.create(
      <Provider store={store}>
        <Display pathname="/test" />
      </Provider>,
    );
    await wait(() => {
      expect(component.toJSON()).toMatchSnapshot();
    });
  });
});
