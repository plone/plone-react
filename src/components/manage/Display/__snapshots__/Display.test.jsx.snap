// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display renders an actions component 1`] = `
Array [
  <label
    htmlFor="display-select"
  >
    View
  </label>,
  <div
    className="react-select-container css-2b097c-container"
    onKeyDown={[Function]}
  >
    <div
      className="react-select__control css-yjz7n5-control"
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
    >
      <div
        className="react-select__value-container react-select__value-container--has-value css-1hwfws3"
      >
        <div
          className="react-select__single-value css-1uccc91-singleValue"
        >
          Summary view
        </div>
        <input
          className="css-62g3xt-dummyInput"
          disabled={false}
          id="react-select-2-input"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          readOnly={true}
          tabIndex="0"
          value=""
        />
      </div>
      <div
        className="react-select__indicators css-1hb7zxy-IndicatorsContainer"
      >
        <span
          className="react-select__indicator-separator css-6fp5c8-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          className="react-select__indicator react-select__dropdown-indicator css-tlfecz-indicatorContainer"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <svg
            className="icon"
            dangerouslySetInnerHTML={
              Object {
                "__html": undefined,
              }
            }
            onClick={null}
            style={
              Object {
                "fill": "#007bc1",
                "height": "24px",
                "width": "auto",
              }
            }
            viewBox=""
            xmlns=""
          />
        </div>
      </div>
    </div>
    <input
      name="display-select"
      type="hidden"
      value="summary_view"
    />
  </div>,
]
`;
